File: mlh_parser2/tests/test_base_email_parsers.py
import pytest
import io
from mlh_parser.email_reader import decode_mail, get_body, get_headers
from .helpers import (
    list_files_with_extension,
    map_to_file_extensions,
    parse_headers_file,
)


directory = "./complete_cases/"
email_files = list_files_with_extension(directory, ".eml")

body_mail_files = [
    map_to_file_extensions(email_f, [".body.pytest"]) for email_f in email_files
]


@pytest.mark.parametrize("email_file, body_file", body_mail_files)
def test_body_parser(email_file, body_file) -> None:
    mail_text = io.open(email_file, mode="rb").read()
    # Read expected body and normalize line endings (CRLF -> LF)
    body = io.open(body_file, mode="r", encoding="utf-8").read().replace("\r\n", "\n")

    mail = decode_mail(mail_text)
    body_from_email = get_body(mail)

    assert body_from_email == body


header_mail_files = [
    map_to_file_extensions(email_f, [".headers.pytest"]) for email_f in email_files
]


@pytest.mark.parametrize("email_file, headers_file", header_mail_files)
def test_header_parser(email_file, headers_file) -> None:
    mail_text = io.open(email_file, mode="rb").read()
    # Parse expected headers from file
    expected_headers = parse_headers_file(headers_file)

    mail = decode_mail(mail_text)
    headers_from_email = get_headers(mail, raw_email=mail_text)

    print("headers_from_email from", headers_from_email.get("from"))
    assert headers_from_email == expected_headers

File: mlh_parser2/tests/test_base_email_parsers.rs
mod common;

use common::{
    list_files_with_extension, map_to_file_extensions, parse_body_file, parse_headers_file,
};
use mlh_parser2::email_reader::{decode_mail, get_body, get_headers};
use std::fs;

#[test]
#[ignore = "fixture exact-match parity with Python needs iteration"]
fn test_body_parser() {
    let directory = "./complete_cases/";
    let email_files = list_files_with_extension(directory, ".eml");

    for email_file in &email_files {
        let fixtures = map_to_file_extensions(email_file, &[".body.pytest"]);
        if fixtures.is_empty() {
            continue;
        }
        let body_file = &fixtures[0];

        if !body_file.exists() {
            continue;
        }

        let mail_bytes = fs::read(email_file).unwrap();
        let expected_body = parse_body_file(body_file);

        let mail = decode_mail(&mail_bytes).unwrap();
        let actual_body = get_body(&mail);

        assert_eq!(
            actual_body, expected_body,
            "Body mismatch for {:?}",
            email_file
        );
    }
}

#[test]
#[ignore = "fixture exact-match parity with Python needs iteration"]
fn test_header_parser() {
    let directory = "./complete_cases/";
    let email_files = list_files_with_extension(directory, ".eml");

    for email_file in &email_files {
        let fixtures = map_to_file_extensions(email_file, &[".headers.pytest"]);
        if fixtures.is_empty() {
            continue;
        }
        let headers_file = &fixtures[0];

        if !headers_file.exists() {
            continue;
        }

        let mail_bytes = fs::read(email_file).unwrap();
        let expected_headers = parse_headers_file(headers_file);

        let mail = decode_mail(&mail_bytes).unwrap();
        let actual_headers = get_headers(&mail);

        for (key, expected_value) in &expected_headers {
            let actual_value = actual_headers.get(key).cloned().unwrap_or_default();
            assert_eq!(
                actual_value, *expected_value,
                "Header mismatch for '{}' in {:?}",
                key, email_file
            );
        }
    }
}
